src/OrbitGl/ScopeTree.cpp

```diff
@@ -0,0 +1,181 @@
+    ordered_nodes_by_depth_[previous_depth].erase(node_timestamp);
+    node->SetDepth(new_depth);
+  }
+
+  // Recurse before inserting the node at new depth to prevent overwriting a child.
+  for (auto& [unused_timestamp, child_node] : node->GetChildren()) {
+    UpdateDepthInSubtree(child_node, new_depth + 1);
+  }
+
+  // Add node to new depth track.
+  ordered_nodes_by_depth_[new_depth].insert({node_timestamp, node});
+}
+
+size_t ScopeTree::CountOrderedNodes() const {
+  size_t count_from_depth = 0;
+  for (auto& pair : ordered_nodes_by_depth_) {
+    count_from_depth += pair.second.size();
+  }
+  return count_from_depth;
+}
+
+void ScopeTree::Print() const {
+  LOG("Printing %u nodes height=%u:", size_, Height());
+  root_->Print();
+  LOG("");
+
+  for (auto& [depth, ordered_nodes] : ordered_nodes_by_depth_) {
+    for (auto& [unused_timestamp, node] : ordered_nodes) {
+      LOG("%u: node_depth:%u id:%u", depth, node->Depth(), node->Id());
+    }
+  }
+}
+
+void ScopeNode::Print(const ScopeNode* node, uint64_t start_time, uint32_t depth) {
+  LOG("d%u %s%lu ScopeNode(%u) %p [%lu, %lu]", node->Depth(), std::string(depth, ' '), start_time,
+      node->id_, node, node->text_box_->GetTimerInfo().start(),
+      node->text_box_->GetTimerInfo().end());
+  for (auto& pair : node->children_) {
+    Print(pair.second, pair.first, depth + 1);
+  }
+}
+
+void ScopeNode::FindHeight(const ScopeNode* node, uint32_t* height, uint32_t current_height) {
+  ORBIT_SCOPE_FUNCTION;
+  *height = std::max(*height, current_height);
+  for (auto& pair : node->children_) {
+    FindHeight(pair.second, height, current_height + 1);
+  }
+}
+
+size_t ScopeNode::CountNodesInSubtree() const {
+  size_t num_nodes = 0;
+  CountNodesInSubtree(this, &num_nodes);
+  return num_nodes;
+}
+
+void ScopeNode::CountNodesInSubtree(const ScopeNode* node, size_t* count) {
+  CHECK(count != nullptr);
+  ++(*count);
+  for (const auto [unused_time, child] : node->children_) {
+    CountNodesInSubtree(child, count);
+  }
+}
+
+std::set<const ScopeNode*> ScopeNode::GetAllNodesInSubtree() const {
+  std::set<const ScopeNode*> node_set;
+  GetAllNodesInSubtree(this, &node_set);
+  return node_set;
+}
+
+void ScopeNode::GetAllNodesInSubtree(const ScopeNode* node, std::set<const ScopeNode*>* node_set) {
+  CHECK(node_set != nullptr);
+  node_set->insert(node);
+  for (const auto [unused_time, child] : node->children_) {
+    GetAllNodesInSubtree(child, node_set);
+  }
+}
+
+uint32_t ScopeNode::Height() const {
+  uint32_t max_height = 0;
+  FindHeight(this, &max_height);
+  return max_height;
+}
+
+ScopeNode* ScopeNode::GetLastChildBeforeOrAtTime(uint64_t time) const {
+  // Get first child before or exactly at "time".
+  if (children_.empty()) return nullptr;
+  auto next_node_it = children_.upper_bound(time);
+  if (next_node_it == children_.begin()) return nullptr;
+  return (--next_node_it)->second;
+}
+
+ScopeNode* ScopeNode::FindDeepestParentForNode(const ScopeNode* node) {
```

Comment:
Nit question: Is there a reason, why you are not doing this call recursive anymore? Personally, I'd find it probably a bit more readable (I required 3 attempts to fully grasp this method -- but still this method is nice and short, so still totally fine).